libs/langchain/langchain/chat_models/huggingface_llama2.py

```diff
@@ -0,0 +1,162 @@
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+from transformers import StoppingCriteria, StoppingCriteriaList
+from transformers.pipelines import TextGenerationPipeline
+
+from langchain.callbacks.manager import (
+    CallbackManagerForLLMRun,
+)
+from langchain.chat_models.base import BaseChatModel
+from langchain.pydantic_v1 import root_validator
+from langchain.schema import ChatResult
+from langchain.schema.messages import (
+    AIMessage,
+    BaseMessage,
+    HumanMessage,
+    SystemMessage,
+)
+from langchain.schema.output import ChatGeneration
+
+B_INST, E_INST = "[INST]", "[/INST]"
+B_SYS, E_SYS = "<<SYS>>", "<</SYS>>"
+
+
+class StoppingCriteriaSub(StoppingCriteria):
+    """Subclass of StoppingCriteria to allow for custom stopping criteria"""
+
+    def __init__(
+        self,
+        stops: Optional[List[torch.Tensor]] = None,
+        device: Union[torch.device, str, None] = None,
+    ):
+        super().__init__()
+        stops = stops or []
+        if device:
+            self.stops = [stop.to(device) for stop in stops]
+        else:
+            self.stops = stops
+
+    def __call__(
+        self,
+        input_ids: torch.LongTensor,
+        scores: torch.FloatTensor,
+        **kwargs: Dict,
+    ) -> bool:
+        for stop_id in self.stops:
+            if (input_ids[0][-torch.numel(stop_id) :] == stop_id).all():
```

Comment:
Do you need to use torch.numel here? It seems as overklill to require torch for this operation